### DISCUSSION TITLE
h/w float issues on Banana Pi BPI-F3 with riscv64 h/w float support

### DISCUSSION THREAD

**Discussion by soccercheng** [pocl/pocl#1540] [Q&A]
Hi,

I'm porting POCL onto [Banana Pi BPI-F3](https://docs.banana-pi.org/en/BPI-F3/BananaPi_BPI-F3), which has riscv64 processors with h/w float/double support.

I am able to successfully build POCL, with some patches of course, on Ubuntu 22.04/X86 with riscv64 cross gcc and llvm-18 compilers, and I try to run examples/vecadd on target device with the following  (s/w float) error: 

[2024-08-26 08:54:42.616333706] PoCL: in fn llvm_codegen at line 200:
 |      LLVM |  written /home/xxx000/.cache/pocl/kcache/tempfile_TxadNq.so.o size 34840
[2024-08-26 08:54:42.616601715] PoCL: in fn llvm_codegen at line 213:
 |      LLVM |  Temporary kernel.so file for kernel vecadd : /home/xxx000/.cache/pocl/kcache/tempfile_ZUG5dh.so
[2024-08-26 08:54:42.616656675] PoCL: in fn llvm_codegen at line 217:
 |   GENERAL |  Linking final module
[2024-08-26 08:54:42.813979385] PoCL: in fn pocl_check_kernel_disk_cache at line 1024:
 |   GENERAL |  Built a specialized WG function: /home/xxx000/.cache/pocl/kcache/HO/PNDKKENNMPHHPNEGDOCHPOLONBHCFABOFECJC/vecadd/128-1-1-goffs0-smallgrid/vecadd.so
dlopen("/home/xxx000/.cache/pocl/kcache/HO/PNDKKENNMPHHPNEGDOCHPOLONBHCFABOFECJC/vecadd/128-1-1-goffs0-smallgrid/vecadd.so") failed with '/home/xxx000/.cache/pocl/kcache/HO/PNDKKENNMPHHPNEGDOCHPOLONBHCFABOFECJC/vecadd/128-1-1-goffs0-smallgrid/vecadd.so: undefined symbol: __addsf3'.
note: missing symbols in the kernel binary might be reported as 'file not found' errors.
Aborted

I've found that the object code, generated from pocl_llvm_codegen of lib/CL/pocl_llvm_wg.cc, contains the "__addsf3' symbol, which is not expected for the platform that supports h/w float/double,

 $ objdump -tC /home/xxx000/.cache/pocl/kcache/HO/PNDKKENNMPHHPNEGDOCHPOLONBHCFABOFECJC/vecadd/128-1-1-goffs0-smallgrid/vecadd.so.o | grep __addsf3
0000000000000000         *UND*	0000000000000000 __addsf3

Can you please instruct me how I can diagnose this issue ?

**Comment by pjaaskel** [comment]
Hi, 

This symbol is missing from the "work group function" which is generated by PoCL via LLVM IR. If you have HW FP you should ensure the work-group functions are built with correct flags that do not emit FP emulation, thus correct machine attributes enabled.  Perhaps this is a lead: https://llvm.org/doxygen/classllvm_1_1TargetLibraryInfo.html#abbec53ddcfefcaaecad3227ecf1335e9

P.S. The PoCL community would appreciate if you upstream any of your fixes to avoid rework for this platform in the future.

**Comment by franz** [comment]
Personally i have only compiled RISC-V directly on a RISC-V machine (no cross-compiling). The main problem i ran into was very similar, i've documented it in `doc/sphinx/source/install.rst`:

```
* on RISC-V, PoCL additionally needs to pass a target ABI flag to the compiler. There is
  some autodetection in PoCL but right now it's limited, and Clang unfortunately does not
  always get the defaults correctly. If you get errors similar to:

      "can't link double-float modules with soft-float modules"

  from linker, then most likely PoCL used the incorrect ABI. You can explicitly
  specify the ABI to use with the HOST_CPU_TARGET_ABI CMake option.
```

You can verify which ABI was autodetected by PoCL by running e.g. `POCL_DEBUG=llvm ./examples/vecadd/vecadd` and looking for `all build options: .....` - in that long string there should be `-target-abi` option.

**Comment by soccercheng** [comment]
I've found where the problem is, but I'm not able to find the root cause due to the complexity of llvm on this part. Below is what I've found:

[pocl_llvm_codegen in pocl/lib/CL/pocl_llvm_wg.cc]

```
1499   if (LLVMGeneratesObjectFiles) {
1500     POCL_MSG_PRINT_LLVM("Generating an object file directly.\n");
1501 #ifdef DUMP_LLVM_PASS_TIMINGS
1502     llvm::TimePassesIsEnabled = true;
1503 #endif
1504     std::cerr << "888888888888888888888888888" << std::endl;
1505     PMObj.run(*Input);
1506     std::cerr << "9999999999999999999999999999" << std::endl;
1507 #ifdef DUMP_LLVM_PASS_TIMINGS
1508     llvm::reportAndResetTimings();
1509 #endif
1510     auto O = SOS.str(); // flush
1511     const char *Cstr = O.data();
1512     size_t S = O.size();
1513     *Output = (char *)malloc(S);
1514     *OutputSize = S;
1515     memcpy(*Output, Cstr, S);
1516     return 0;
1517   }
```
PMObj.run calls "RISCVSubtarget::initializeSubtargetDependencies" twice, **but I don't know the reason why ?**   

[llvm/lib/Target/RISCV/RISCVSubtarget.cpp]

 ```
 72 RISCVSubtarget &
 73 RISCVSubtarget::initializeSubtargetDependencies(const Triple &TT, StringRef CPU,
 74                                                 StringRef TuneCPU, StringRef FS,
 75                                                 StringRef ABIName) {
 76   // Determine default and user-specified characteristics
 77   bool Is64Bit = TT.isArch64Bit();
 78   if (CPU.empty() || CPU == "generic")
 79     CPU = Is64Bit ? "generic-rv64" : "generic-rv32";
 80 
 81   if (TuneCPU.empty())
 82     TuneCPU = CPU;
 83 
 84   { 
 85     errs() << "\n\nCPU: " + CPU +
 86               "\nTuneCPU: " + TuneCPU +
 87               "\nFS: " + FS + 
 88               "\nABIName: " + ABIName +
 89               "\n\n";
 90   }
```
I've tried to dump the argument accordingly, but the FS  of 1st call is different from 2nd call

[first time]
    CPU: generic-rv64
    TuneCPU: generic-rv64
    FS: +64bit
   ABIName: lp64d

[second time]
    CPU: generic-rv64
    TuneCPU: generic-rv64
    FS: +64bit,+a,+c,+d,+f,+m,+relax,+v,+zicsr,+zifencei, ....
    ABIName: lp64d

For the first PASS, FS does NOT contain "+d" and "+f", this is the reason why POCL/LLVM does generate h/w floating point instruction accordingly, I don't know why 1st and 2nd passes use different features and I'll keep on looking into this issue.

**Reply by pjaaskel** [reply]
Are you working on the latest main checkout? I committed last week a fix (related to passing machine info in the second phase of opts) that might help here as well.

**Reply by soccercheng** [reply]
Would you mind if you can provide the commit ID for my reference ?

**Reply by pjaaskel** [reply]
https://github.com/pocl/pocl/pull/1535/commits/19dc70bbe927a6f6210ccd45b6a4c2a49fe6da96

**Comment by soccercheng** [comment]
In order to make this work, I need to add features to target @ these 2 places, But I am not sure if these are the right place/time to add features to target ?

[lib/CL/pocl_llvm_wg.cc]

```
static TargetMachine *GetTargetMachine(cl_device_id device) {
  ....
  // OpenASIP targets are not in the registry
  if (!TheTarget) {
    return nullptr;
  }

  const char *targetFeatures = "";
  targetFeatures = "+64bit,+a,+c,+d,+f,+m,+relax,+v,+zicsr,+zifencei,+zve32f,+zve32x,+zve64d,+zve64f,+zve64x,"
                   "+zvl128b,+zvl256b,+zvl32b,+zvl64b";
  TargetMachine *TM = TheTarget->createTargetMachine(
      DevTriple.getTriple(), MCPU, StringRef(targetFeatures), TargetOptions(),
      Reloc::PIC_, CodeModel::Small,
#if LLVM_MAJOR >= 18
      CodeGenOptLevel::Aggressive);
#else
      CodeGenOpt::Aggressive);
#endif

  ....

  return TM;
}
```

[lib/CL/pocl_llvm_build.cc] 

```
int pocl_llvm_build_program(cl_program program,
                            unsigned device_i,
                            cl_uint num_input_headers,
                            const cl_program *input_headers,
                            const char **header_include_names,
                            int linking_program)
{
  ....

  clang::TargetOptions &ta = pocl_build.getTargetOpts();
  ta.Triple = device->llvm_target_triplet;
  if (device->llvm_cpu != NULL)
    ta.CPU = device->llvm_cpu;

  if(1)    {    
    auto ISAInfo = llvm::RISCVISAInfo::parseArchString(device->llvm_arch /* rv64imafdcvzicsr_zvl256b_zifencei */,
                                                       true /* EnableExperimentalExtensions */);
    for (const std::string &Feature : (*ISAInfo)->toFeatures(/*AddAllExtension=*/true,
                                                             /*IgnoreUnknown=*/false))    {    
        std::cout << "### Add Feature: " << Feature << std::endl;
        const char *ft = Feature.c_str();
        ta.Features.push_back(&(ft[1]));
        if(*ft == '+') 
            ta.FeatureMap[&(ft[1])] = true;
        else
            ta.FeatureMap[&(ft[1])] = false;
    }

    bool EnableExperimentalExtensions = true;
    ta.Features.push_back("experimental");
    ta.FeatureMap["experimental"] = EnableExperimentalExtensions;
  }

#ifdef DEBUG_POCL_LLVM_API
  std::cout << "### Triple: " << ta.Triple.c_str() << ", CPU: " << ta.CPU.c_str() << std::endl
            << "    TuneCPU:" << ta.TuneCPU.c_str() << ", ABI: " << ta.ABI.c_str() << std::endl;
#endif

.....
}
```

**Reply by pjaaskel** [reply]
Perhaps they are right places, but you cannot unconditionally assume we are always targeting a RISC-V target you have (or any RISC-V target, your code likely breaks other ISAs). You should check from the triple that it's a riscv target and then somehow detect the features you can enable.  Not sure how it's the best to autodetect those features though without hard coding - I'm afraid you need to research it a bit. Is there cpuinfo infra in LLVM for the detection or such?

**Reply by soccercheng** [reply]
Yes, many thanks for your suggestions, I've adjust the code like the following:

```
#if defined(__riscv)
  if(NULL != device->llvm_arch)    {    
    POCL_MSG_PRINT_LLVM("LLVM Aech (-march): %s\n", device->llvm_arch);
    auto ISAInfo = llvm::RISCVISAInfo::parseArchString(device->llvm_arch,
                                                       true, /* EnableExperimentalExtensions */
                                                       true, /* ExperimentalExtensionVersionCheck */
                                                       true  /* IgnoreUnknown */);
    for (const std::string &Feature : (*ISAInfo)->toFeatures(true, /* AddAllExtension */
                                                             true  /* IgnoreUnknown */))    {    
      const char *ft = Feature.c_str();
      ta.Features.push_back(&(ft[1]));
      if(*ft == '+') 
        ta.FeatureMap[&(ft[1])] = true;
      else 
        ta.FeatureMap[&(ft[1])] = false;
    }    

    bool EnableExperimentalExtensions = true;
    ta.Features.push_back("experimental");
    ta.FeatureMap["experimental"] = EnableExperimentalExtensions;
  }
#endif
```

As the value for "device->llvm_arch",
it is queried from devicetree(/sys/firmware/devicetree/base/cpus/cpu@0/riscv,isa" or cpuinfo(/proc/cpuinfo).

After making the adjustment, the code seems work fine on my Banana Pi.

**Reply by pjaaskel** [reply]
A cleaner way would be to introduce an llvm_features or similar string variable to the device API and populate it in pthread/basic cpuinfo() which can be then used here.  This we could accept to the PoCL upstream project if you send a pull request and would like to contribute to the open source project. Thanks!
